### DISCUSSION TITLE
SHA-256 repository support

### DISCUSSION THREAD

**Discussion by bk2204** [libgit2/libgit2#5840] [Ideas]
I'm looking into adding SHA-256 repository support for libgit2.  I have [a branch with some starting points](https://github.com/bk2204/libgit2/tree/sha256).

One thing that's become very obvious to me is that in all likelihood, this is going to be an ABI-breaking change, since in many cases (e.g., the index) we allow users to pass in arbitrary files from the system without a repository attached.  Git generally doesn't mark its objects with the hash algorithm, so when we're parsing formats we have no way of knowing what the hash algorithm is without being told.

I assume this isn't a problem, but since I have several days to work on this project and it's possible a PR will come in during that time, I expect it's a thing y'all will want to be aware of.

**Comment by ethomson** [comment]
> One thing that's become very obvious to me is that in all likelihood, this is going to be an ABI-breaking change

Yes, it's hard for me to imagine how we could do this without an ABI change.  But it's also hard for me to imagine how we do this without an _API_ change, so I'm much less concerned about the ABI changes in comparison.

I think that one initial step that would be helpful for me is to understand the status of SHA-256 support in git itself.  We're at the stage where we're ["[introducing] the the objectFormat extension (without compatObjectFormat)"](https://github.com/git/git/blob/master/Documentation/technical/hash-function-transition.txt#L644).  Is that roughly accurate?

I should be thinking about this with the ultimate goal being that `compatObjectFormat` is a thing?  It's a little hard for me to reason about what libgit2's API should look like without being able to really try it out in anger in git, but I just want to make sure that I still understand the desired end state for git.

**Reply by bk2204** [reply]
Yes, that's roughly accurate.  Git has support for the stage 4 implementation, but no other stages; that is, you can have a Git repository with only SHA-1 or only SHA-256, but they do not interoperate, nor are different hashes currently used in the same repo.

The goal is that `compatObjectFormat` is a thing.  The first steps in Git, which I am working on now, are to get objects such that the loose object index and pack index v3 are written with both algorithms, and then we will progress to fetch and push interoperability, which will be the first time that `compatObjectFormat` is used by users.  Beyond that point, we'll get Git to handle different algorithms on input and output, which may or may not be relevant for libgit2.

I definitely think an API change is coming.  For example, we'll need to make `git_oid` take a hash algorithm member so we can format it properly as a string.  I have some code to do that, but at the moment I'm trying to avoid pushing up things which fail the testsuite, although I may have to do so, depending on how much time I end up having to work on this.

**Comment by jimboolio** [comment]
It should probably be mentioned bit more clearly in readme/docs that there currently isn't support for sha256. For me it was not a problem to migrate my repository back to sha1 losing my commit history, but for someone else that might be. And at least the downstream project I'm using did not report that the failures were caused by this, so debugging and figuring out the dependencies to here took some time.

**Comment by ethomson** [comment]
@jimboolio If you build libgit2 with `-DEXPERIMENTAL_SHA256` then we have SHA256. It's somewhat well-tested, but not _battle tested in a Git forge_ yet. It's marked "experimental" as I expect the API surface will change somewhat as git evolves their SHA256 support.

**Comment by ethomson** [comment]
But, I agree, the state of SHA256 support should be mentioned more prominently. 😁
